### PR TITLE
[With interface workaround] Auth Web OTP Retrieval 

### PR DESCRIPTION
**Pull request by Evanition**
This branch contained all the changes related to webOTP Autofill.
Includes changes in 
- confirmWithWebOTP
- overloaded verifyPhoneNumber/_verifyPhoneNumber
- overloaded signInWithPhoneNumber
- interfaces for webOTP typescript support
- unit tests

### CODE REVIEW

File: packages/auth/src/core/errors.ts [hunk 1]
```diff
@@ -381,7 +381,8 @@ function _debugErrorMap(): ErrorMap<AuthErrorCode> {
       'The reCAPTCHA version is missing when sending request to the backend.',
     [AuthErrorCode.INVALID_REQ_TYPE]: 'Invalid request parameters.',
     [AuthErrorCode.INVALID_RECAPTCHA_VERSION]:
-      'The reCAPTCHA version is invalid when sending request to the backend.'
+      'The reCAPTCHA version is invalid when sending request to the backend.',
```
Comment (kevinthecheung) [inline packages/auth/src/core/errors.ts:384]:
```suggestion
      'The reCAPTCHA version sent to the backend is invalid.',
```

File: packages/auth/src/model/public_types.ts [hunk 2]
```diff
@@ -587,6 +587,18 @@ export interface ConfirmationResult {
...
    */
   confirm(verificationCode: string): Promise<UserCredential>;
+  /**
+   *
+   * Automatically fetches the verification code from SMS message. Then, it calls confirm(verificationCode) to finish a phone number sign-in, link, or reauthentication.
+   *
+   * @param auth - the current {@link Auth} instance
+   * @param webOTPTimeout - Error would be thrown if WebOTP does not resolve within this specified timeout parameter (milliseconds).
+   *
+   */
```
Comment (kevinthecheung) [inline packages/auth/src/model/public_types.ts:597]:
```suggestion
  /**
   *
   * Automatically fetches a verification code from an SMS message. Then, calls confirm(verificationCode) to finish a phone number sign-in, link, or reauthentication.
   *
   * @param auth - the current {@link Auth} instance
   * @param webOTPTimeout - Error would be thrown if WebOTP does not resolve within this specified timeout parameter (milliseconds).
   *
   */
```

File: packages/auth/src/platform_browser/providers/phone.ts [hunk 3]
```diff
@@ -101,11 +101,64 @@ export class PhoneAuthProvider {
...
+   * ```
+   *
+   * @param phoneInfoOptions - The user's {@link PhoneInfoOptions}. The phone number should be in
+   * E.164 format (e.g. +16505550101).
+   * @param applicationVerifier - For abuse prevention, this method also requires a
+   * {@link ApplicationVerifier}. This SDK includes a reCAPTCHA-based implementation,
+   * {@link RecaptchaVerifier}.
+   *
+   * @returns A Promise for a UserCredential.
+   */
```
Comment (kevinthecheung) [inline packages/auth/src/platform_browser/providers/phone.ts:136]:
```suggestion
  /**
   *
   * Completes the phone number authentication flow by sending a verification code to the
   * given phone number, automatically retrieving the verification code from the SMS message,
   * and signing the user in.
   *
   * @example
   * ```javascript
   * const provider = new PhoneAuthProvider(auth);
   * const userCredential = await provider.verifyPhoneNumber(phoneNumber, applicationVerifier, 10);
   * ```
   *
   * @example
   * An alternative flow is provided using the `signInWithPhoneNumber` method.
   * ```javascript
   * const userCredential = signInWithPhoneNumber(auth, phoneNumber, applicationVerifier, 10);
   * ```
   *
   * @param phoneInfoOptions - The user's {@link PhoneInfoOptions}. The phone number should be in
   * E.164 format (for example, +16505550101).
   * @param applicationVerifier - For abuse prevention, this method also requires an
   * {@link ApplicationVerifier}. This SDK includes a reCAPTCHA-based implementation,
   * {@link RecaptchaVerifier}.
   *
   * @returns A Promise for a UserCredential.
   */
```

**Review by kevinthecheung — APPROVED** [review]

File: packages/auth/src/platform_browser/providers/phone.ts [hunk 4]
```diff
@@ -101,11 +101,64 @@ export class PhoneAuthProvider {
...
+   * ```javascript
+   * const userCredential = signInWithPhoneNumber(auth, phoneNumber, applicationVerifier, 10);
+   * ```
+   *
+   * @param phoneInfoOptions - The user's {@link PhoneInfoOptions}. The phone number should be in
+   * E.164 format (e.g. +16505550101).
+   * @param applicationVerifier - For abuse prevention, this method also requires a
+   * {@link ApplicationVerifier}. This SDK includes a reCAPTCHA-based implementation,
+   * {@link RecaptchaVerifier}.
+   *
```
Comment (prameshj) [inline packages/auth/src/platform_browser/providers/phone.ts:134]: We should document the timeout param also.

File: packages/auth/src/platform_browser/providers/phone.ts [hunk 5]
```diff
@@ -101,11 +101,64 @@ export class PhoneAuthProvider {
...
+    phoneOptions: PhoneInfoOptions | string,
+    applicationVerifier: ApplicationVerifier,
+    webOTPTimeout: number
+  ): Promise<UserCredential>;
+
+  verifyPhoneNumber(
+    phoneOptions: PhoneInfoOptions | string,
+    applicationVerifier: ApplicationVerifier,
+    webOTPTimeout?: number
+  ): Promise<unknown> {
```
Comment (prameshj) [inline packages/auth/src/platform_browser/providers/phone.ts:147]: nit: can we explore if "Promise<UserCredential| string>" will work?

File: packages/auth/src/platform_browser/strategies/phone.test.ts [hunk 6]
```diff
@@ -47,6 +47,11 @@ import {
 use(chaiAsPromised);
 use(sinonChai);
 
+//interfaces added to provide typescript support for webopt autofill
```
Comment (prameshj) [inline packages/auth/src/platform_browser/strategies/phone.test.ts:50]: Is this needed in the test fil?

File: packages/auth/src/platform_browser/strategies/phone.ts [hunk 7]
```diff
@@ -65,14 +78,92 @@ class ConfirmationResultImpl implements ConfirmationResult {
...
+        const myErr = _errorWithCustomMessage(
+          auth,
+          AuthErrorCode.WEB_OTP_NOT_RETRIEVED,
+          `Web OTP code is not fetched before timeout`
+        ) as WebOTPError;
+        myErr.confirmationResult = this;
+        throw myErr;
+      }, webOTPTimeout * 1000);
+
+      // @ts-ignore - ignore types for testing
```
Comment (prameshj) [inline packages/auth/src/platform_browser/strategies/phone.ts:100]: nit: remove this. IIRC, we tried this to avoid adding custom interfaces, but this line didnt help.

File: packages/auth/src/platform_browser/strategies/phone.ts [hunk 8]
```diff
@@ -65,14 +78,92 @@ class ConfirmationResultImpl implements ConfirmationResult {
...
+        .then(async content => {
+          if (
+            content === undefined ||
+            content === null ||
+            content.code === undefined
+          ) {
+            const myErr = _errorWithCustomMessage(
+              auth,
+              AuthErrorCode.WEB_OTP_NOT_RETRIEVED,
+              `Web OTP code is not valid`
```
Comment (prameshj) [inline packages/auth/src/platform_browser/strategies/phone.ts:119]: Should be a different error message than "code is not valid"?

File: packages/auth/src/platform_browser/strategies/phone.ts [hunk 9]
```diff
@@ -103,16 +194,75 @@ export async function signInWithPhoneNumber(
...
+ * ```javascript
+ * // 'recaptcha-container' is the ID of an element in the DOM.
+ * const applicationVerifier = new firebase.auth.RecaptchaVerifier('recaptcha-container');
+ * const userCredential = await signInWithPhoneNumber(auth, phoneNumber, applicationVerifier, 10);
+ * ```
+ *
+ * @param auth - The {@link Auth} instance.
+ * @param phoneNumber - The user's phone number in E.164 format (e.g. +16505550101).
+ * @param appVerifier - The {@link ApplicationVerifier}.
+ * @param webOTPTimtout - Errors would be thrown if WebOTP autofill is used and does not resolve within this specified timeout parameter (milliseconds).
```
Comment (prameshj) [inline packages/auth/src/platform_browser/strategies/phone.ts:226]: couple of nits:
1) fix spelling - webOTPTimeout
2) the timeout is in seconds, not milliseconds, correct?
3) Maybe word it as "The timeout (in seconds) to wait for WebOTP to autofill the SMS code"?

File: packages/auth/test/integration/flows/totp.test.ts [hunk 10]
```diff
@@ -55,9 +55,10 @@ describe(' Integration tests: Mfa enrollement using totp', () => {
   beforeEach(async () => {
     emulatorUrl = getEmulatorUrl();
     if (!emulatorUrl) {
-      mfaUser = null;
       auth = getTestInstance();
       displayName = 'totp-integration-test';
+      const cr = await signInWithEmailAndPassword(auth, email, password);
```
Comment (prameshj) [inline packages/auth/test/integration/flows/totp.test.ts:60]: Let's move changes in this file to a separate CL (if needed)

File: packages/auth/src/platform_browser/strategies/phone.ts [hunk 11]
```diff
@@ -241,15 +405,31 @@ export async function _verifyPhoneNumber(
...
+          return confirmationRes.confirmWithWebOTP(authInternal, webOTPTimeout);
+        } catch (error) {
+          throw error;
+        }
+      } else {
+        return verificationId;
+      }
     }
+  } catch (error) {
+    console.log('reached error catching block in _verifyphonenumber');
```
Comment (prameshj) [inline packages/auth/src/platform_browser/strategies/phone.ts:431]: nit: remove

**Review by prameshj — COMMENTED** [review]
Thanks for working on this!

File: packages/auth/test/integration/flows/totp.test.ts (hunk 10 above)
Comment (AngelAngelXie) [reply packages/auth/test/integration/flows/totp.test.ts:60]: This has been removed completely. Sorry about the confusion! :)

File: packages/auth/src/platform_browser/providers/phone.ts (hunk 5 above)
Comment (AngelAngelXie) [reply packages/auth/src/platform_browser/providers/phone.ts:149]: Yes. This would also work :) I like that it is more specific than using "unknown" - even though this will not be public for developers to see. I have replaced "unknown" in the return types. Please feel free to comment if you believe "unknown" is preferred.

File: packages/auth/src/platform_browser/strategies/phone.test.ts (hunk 6 above)
Comment (AngelAngelXie) [reply packages/auth/src/platform_browser/strategies/phone.test.ts:50]: Yes, I believe this is needed for us to use OTPCredential in the unit tests. It can be removed once typescript support is added for the WebOTP API.

File: packages/auth/src/platform_browser/strategies/phone.ts (hunk 8 above)
Comment (AngelAngelXie) [reply packages/auth/src/platform_browser/strategies/phone.ts:118]: Would "the atuo-retrieved credential or code is not defined" be a better error message?

File: packages/auth/demo/database.rules.json
Comment (renkelvin) [inline packages/auth/demo/database.rules.json:1]: revert unrelated changes

File: packages/auth/demo/public/404.html
Comment (renkelvin) [inline packages/auth/demo/public/404.html:1]: revert unrelated changes

File: packages/auth/demo/yarn.lock
Comment (renkelvin) [inline packages/auth/demo/yarn.lock:1]: revert yarn.lock changes

File: packages/auth/demo/yes/index.html
Comment (renkelvin) [inline packages/auth/demo/yes/index.html:1]: revert unrelated changes

File: packages/auth/src/api/index.ts [hunk 12]
```diff
@@ -134,7 +134,7 @@ export async function _performApiRequest<T, V>(
     }
 
     return FetchProvider.fetch()(
-      _getFinalTarget(auth, auth.config.apiHost, path, query),
```
Comment (renkelvin) [inline packages/auth/src/api/index.ts:137]: This should only be used locally

File: packages/auth/src/platform_browser/strategies/phone.ts (hunk 8 above)
Comment (renkelvin) [reply packages/auth/src/platform_browser/strategies/phone.ts:118]: Sounds good to me.

File: packages/auth/test/integration/webdriver/static/rollup.config.js
Comment (renkelvin) [inline packages/auth/test/integration/webdriver/static/rollup.config.js:1]: revert unrelated changes

File: packages/auth/test/integration/flows/totp.test.ts
Comment (renkelvin) [inline packages/auth/test/integration/flows/totp.test.ts:1]: revert unrelated changes

File: common/api-review/auth.api.md
Comment (renkelvin) [inline common/api-review/auth.api.md:1]: We should hide the public api changes, otherwise we can't merge this PR.

**Review by renkelvin — APPROVED** [review]
Great, thanks!
